src/modules/flow/lsm9ds0/lsm9ds0.c

```diff
@@ -0,0 +1,510 @@
+
+    op[4].type = SOL_I2C_READ;
+    op[4].reg = OUT_Z_L_G;
+    op[5].type = SOL_I2C_READ;
+    op[5].reg = OUT_Z_H_G;
+}
+
+static void
+_read_gyro_done(void *cb_data, ssize_t status)
+{
+    struct sol_direction_vector gyro;
+    struct lsm9ds0_gyro_data *mdata = cb_data;
+
+    mdata->pending = NULL;
+
+    if (status <= 0) {
+        SOL_ERR("Couldn't read LSM9DS0 Gyroscope.");
+        return;
+    }
+
+    _parse_raw_data(&gyro, sol_vector_get_no_check(&mdata->queue, 0), (mdata->scale / SAMPLE_RES));
+
+    sol_flow_send_direction_vector_packet(mdata->node,
+        SOL_FLOW_NODE_TYPE_LSM9DS0_GYROSCOPE__OUT__OUT, &gyro);
+}
+
+static void
+_init_gyro_done(void *data, ssize_t status)
+{
+    struct lsm9ds0_gyro_data *mdata = data;
+
+    mdata->pending = NULL;
+
+    if (status <= 0) {
+        mdata->init = false;
+        SOL_ERR("Couldn't initialize LSM9DS0 Gyroscope.");
+        return;
+    }
+
+    sol_vector_clear(&mdata->queue);
+    _gyro_read_queue(mdata);
+
+    mdata->init = true;
+}
+
+static int
+lsm9ds0_gyro_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    uint8_t scale_reg = 0x00;
+    struct lsm9ds0_gyro_data *mdata = data;
+    const struct sol_flow_node_type_lsm9ds0_gyroscope_options *opts =
+        (const struct sol_flow_node_type_lsm9ds0_gyroscope_options *)options;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(opts,
+        SOL_FLOW_NODE_TYPE_LSM9DS0_GYROSCOPE_OPTIONS_API_VERSION, -EINVAL);
+
+    mdata->node = node;
+    mdata->init = false;
+
+    mdata->i2c = sol_i2c_open(opts->i2c_bus, SOL_I2C_SPEED_10KBIT);
+    SOL_NULL_CHECK_MSG(mdata->i2c, -EINVAL, "Failed to open i2c bus: %u", opts->i2c_bus);
+
+    mdata->addr = opts->i2c_addr;
+    mdata->scale = opts->scale;
+
+    if (mdata->scale == 2000) {
+        scale_reg = 0x0C;
+    } else if (mdata->scale == 500) {
+        scale_reg = 0x04;
+    } else if (mdata->scale != 245) {
+        SOL_WRN("Invalid Scale option. Using 245 dps. Valid options are: 245, 500, 2000.");
+        mdata->scale = 245;
+    }
+
+    _gyro_init_queue(mdata, scale_reg);
+
+    mdata->pending = sol_i2c_dispatcher_add_op_set(mdata->i2c, mdata->addr, &mdata->queue,
+        _init_gyro_done, mdata, 0);
+
+    if (!mdata->pending) {
+        SOL_ERR("Couldn't initialize LSM9DS0 Gyroscope.");
+        return -1;
+    }
+
+    return 0;
+}
+
+static void
+lsm9ds0_gyro_close(struct sol_flow_node *node, void *data)
+{
+    struct lsm9ds0_gyro_data *mdata = data;
+
+    if (mdata->i2c) {
+        if (mdata->pending)
+            sol_i2c_dispatcher_remove_op_set(mdata->i2c, mdata->pending);
+        sol_i2c_close(mdata->i2c);
+    }
+
+    sol_vector_clear(&mdata->queue);
+}
+
+static int
+lsm9ds0_gyro_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    struct lsm9ds0_gyro_data *mdata = data;
+
+    if (!mdata->init || mdata->pending)
+        return -1;
```

Comment:
the port process errors will be just sent to stderr (or ignored) by the flow engine. If that should have some meaning, then send an error packet.